### DISCUSSION TITLE
evaluating the displacement at the point

### DISCUSSION THREAD

**Discussion by beksar** [firedrakeproject/firedrake#3323] [Firedrake support]
evaluating point displacement while using CG finite element is possible by 
u.at(xlength/2,ylength/2) in 2D. 
When trying to evaluate displacement on a precise point while using higher-order elements such as Hermite and Argyris, It gives error:
got an error: in point_evaluation
    raise NotImplementedError("TODO: not yet thought about it")
NotImplementedError: TODO: not yet thought about it. 
It is most likely because of their complex shape functions. 
How to solve it?

**Comment by colinjcotter** [comment]
With Hermite and Argyris you can find a CG space that contains them (cubic Hermite is contained in CG3 and Argyris in CG5), and interpolate to that first – this just represents the same function in a different basis so there is no error (except for roundoff). Then you can use .at for the interpolated CG function.

From: beksar ***@***.***>
Date: Friday, 12 January 2024 at 17:03
To: firedrakeproject/firedrake ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [firedrakeproject/firedrake] evaluating the displacement at the point (Discussion #3323)

evaluating point displacement while using CG finite element is possible by
u.at(xlength/2,ylength/2) in 2D.
When trying to evaluate displacement on a precise point while using higher-order elements such as Hermite and Argyris, It gives error:
got an error: in point_evaluation
raise NotImplementedError("TODO: not yet thought about it")
NotImplementedError: TODO: not yet thought about it.
It is most likely because of their complex shape functions.
How to solve it?

**Comment by beksar** [comment]
Pasted a small part of the code to clarify
V = VectorFunctionSpace(mesh, FE, 5 ) 
.
.
#Solve
Gain = derivative(Form,u,du)
problem = NonlinearVariationalProblem(Form, u, bcs=bc, J=Gain)
solver  = NonlinearVariationalSolver(problem)
solver.parameters.update(snes_solver_parameters)
solver.solve()

projected = File("u_2D.pvd", project_output=True)
projected.write(u)

How to interpolate the way you say?

**Reply by ksagiyam** [reply]
Minimal working example would be:
```
from firedrake import *
  

mesh = UnitSquareMesh(2, 2)
x, y = SpatialCoordinate(mesh)
degree = 5
V = FunctionSpace(mesh, "Argyris", degree)
Vcg = FunctionSpace(mesh, "P", degree)
f = Function(V).project((x + y) ** degree, solver_parameters={"ksp_type": "preonly",
                                                              "pc_type": "lu"})
fcg = Function(Vcg).interpolate(f)
assert abs(assemble((fcg - f) ** 2 * dx)) < 1.e-15
assert abs(fcg.at([.3, .4]) - (.3 + .4) ** degree) < 1.e-15
```

**Reply by beksar** [reply]
Thanks for your help, solved the problem in this way, 

.
..
...
Vcg = VectorFunctionSpace(mesh, "P", degree)
u_interpolated = Function(Vcg)

# Project the values from u(solved) to u_interpolated
project(u, u_interpolated)
u_interpolated.at(0,0) # in 2D
